cylc/uiserver/websockets/template.py

```diff
@@ -0,0 +1,130 @@
+    function locationQuery(params, location) {
+      return (location ? location: '') + '?' + Object.keys(params).map(function (key) {
+        return encodeURIComponent(key) + '=' +
+          encodeURIComponent(params[key]);
+      }).join('&');
+    }
+    // Derive a fetch URL from the current URL, sans the GraphQL parameters.
+    var graphqlParamNames = {
+      query: true,
+      variables: true,
+      operationName: true
+    };
+    var otherParams = {};
+    for (var k in parameters) {
+      if (parameters.hasOwnProperty(k) && graphqlParamNames[k] !== true) {
+        otherParams[k] = parameters[k];
+      }
+    }
+    var fetcher;
+    if (true) {
+      var subscriptionsClient = new window.SubscriptionsTransportWs.SubscriptionClient('${subscriptionsEndpoint}', {
+        reconnect: true
+      });
+      fetcher = window.GraphiQLSubscriptionsFetcher.graphQLFetcher(subscriptionsClient, graphQLFetcher);
+    } else {
+      fetcher = graphQLFetcher;
+    }
+    // We don't use safe-serialize for location, because it's not client input.
+    var fetchURL = locationQuery(otherParams, '${endpointURL}');
+    // Defines a GraphQL fetcher using the fetch API.
+    function graphQLFetcher(graphQLParams) {
+        return fetch(fetchURL, {
+          method: 'post',
+          headers: {
+            'Accept': 'application/json',
+            'Content-Type': 'application/json',
+          },
+          body: JSON.stringify(graphQLParams),
+          credentials: 'include',
+        }).then(function (response) {
+          return response.text();
+        }).then(function (responseBody) {
+          try {
+            return JSON.parse(responseBody);
+          } catch (error) {
+            return responseBody;
+          }
+        });
+    }
+    // When the query and variables string is edited, update the URL bar so
+    // that it can be easily shared.
+    function onEditQuery(newQuery) {
+      parameters.query = newQuery;
+      updateURL();
+    }
+    function onEditVariables(newVariables) {
+      parameters.variables = newVariables;
+      updateURL();
+    }
+    function onEditOperationName(newOperationName) {
+      parameters.operationName = newOperationName;
+      updateURL();
+    }
+    function updateURL() {
+      history.replaceState(null, null, locationQuery(parameters) + window.location.hash);
+    }
+    // Render <GraphiQL /> into the body.
+    ReactDOM.render(
+      React.createElement(GraphiQL, {
+        fetcher: fetcher,
+        onEditQuery: onEditQuery,
+        onEditVariables: onEditVariables,
+        onEditOperationName: onEditOperationName,
+      }),
+      document.body
+    );
+  </script>
+</body>
+</html>''').substitute(
+        GRAPHIQL_VERSION='0.10.2',
+        SUBSCRIPTIONS_TRANSPORT_VERSION='0.7.0',
+        subscriptionsEndpoint=f'ws://localhost:8000/{base_url}subscriptions',
+        # subscriptionsEndpoint='ws://localhost:8000/',
```

Comment:
This doesn't work when you deploy JupyterHub to a different port! :angry: 

Realized that I couldn't use GraphiQL when deployed it behind a Nginx rev proxy (following JupyterHub documentation). We can fix it later maybe... though I am not sure how this could be fixed in this template...

Reply:
Maybe turn it into a Jinja2 (or other) template? then the handler can render?

Reply:
Ah.. It's a python script.. Maybe directly via handler?

Reply:
We had the same line of thinking! I started by thinking that I could template'ize.. then thought about doing it in Python... then JS... and then I remember that I actually tried it, but I think it failed because - for some reason - the relative path (i.e.  or `/subscriptions`) or the abbreviated form `ws://:8000/subscriptions` did not work in my browser.

And I couldn't quickly find a way to inject the current server name. Gotta have another go with more calm/time over the next week :+1: thanks!